### PR TITLE
New elasticsearch data loader

### PR DESCRIPTION
**Pull request by tony-sappe**
`es_rapidloader` is a new loading script which uses multiprocessing to download Postgres data and upload to Elasticsearch in separate processes.

## This script relies on an index template for the index datatypes and other mapping info

Modified `es_configure` to create the new template and apply desired settings

### CODE REVIEW

File: usaspending_api/etl/es_etl_helpers.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+    printf({
+        'msg': 'Elasticsearch Index loading took {}s'.format(perf_counter() - start),
+        'job': job.name,
+        'f': 'ES Ingest'
+    })
+
+
+def deleted_transactions(client, config):
+    deleted_ids = gather_deleted_ids(config)
+    id_list = [{'key': c, 'col': UNIVERSAL_TRANSACTION_ID_NAME} for c in deleted_ids]
```
Comment (dpb-bah) [inline usaspending_api/etl/es_etl_helpers.py]: Can you rename `c` to `deleted_id` here for readability?

File: usaspending_api/etl/es_etl_helpers.py [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+def chunks(l, n):
+    """Yield successive n-sized chunks from l."""
+    for i in range(0, len(l), n):
+        yield l[i:i + n]
+
+
+def delete_transactions_from_es(client, id_list, job_id, config, index=None, size=50000):
+    '''
+    id_list = [{key:'key1',col:'tranaction_id'},
+               {key:'key2',col:'generated_unique_transaction_id'}],
```
Comment (dpb-bah) [inline usaspending_api/etl/es_etl_helpers.py]: Col is currently hardcoded to be `generated_unique_transaction_id`, will it ever be `tranaction_id`

File: usaspending_api/etl/es_etl_helpers.py [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+            job.count = download_csv(count_sql, copy_sql, job.csv, job.name, config['verbose'])
+            done_jobs.put(job)
+            printf({
+                'msg': 'CSV "{}" copy took {} seconds'.format(job.csv, perf_counter() - start),
+                'job': job.name,
+                'f': 'Download'
+            })
+            sleep(1)
+
+    done_jobs.put(DataJob(None, None, None, None, None))
```
Comment (dpb-bah) [inline usaspending_api/etl/es_etl_helpers.py]: Can you insert a comment here to explain this?

File: usaspending_api/etl/es_etl_helpers.py [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+import tempfile
+
+from collections import defaultdict
+from datetime import datetime
+from django.db import connection
+from elasticsearch import helpers
+from time import perf_counter, sleep
+
+AWARD_DESC_CATEGORIES = {
+    'loans': 'loans',
```
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/etl/es_etl_helpers.py]: the keys are the terms currently used in postgres DB. We can't alter those :)

File: usaspending_api/etl/es_etl_helpers.py (hunk 3 above)
Comment (tony-sappe) [reply usaspending_api/etl/es_etl_helpers.py]: Will do

File: usaspending_api/etl/es_etl_helpers.py (hunk 1 above)
Comment (tony-sappe) [reply usaspending_api/etl/es_etl_helpers.py]: yes

File: usaspending_api/etl/es_etl_helpers.py [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+
+        for uid in new_ids:
+            if uid in deleted_ids:
+                if deleted_ids[uid]['timestamp'] < obj.last_modified:
+                    deleted_ids[uid]['timestamp'] = obj.last_modified
+            else:
+                deleted_ids[uid] = {'timestamp': obj.last_modified}
+
+    if config['verbose']:
+        for k, v in deleted_ids.items():
```
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/etl/es_etl_helpers.py]: yes

File: usaspending_api/etl/es_etl_helpers.py (hunk 2 above)
Comment (tony-sappe) [reply usaspending_api/etl/es_etl_helpers.py]: `transaction_id` doesn't work for the records of deleted transactions from the broker stored in S3. While the code ((should)) be able to handle any column, the goal is to use `generated_unique_transaction_id` everywhere. Early tests show that if this field is loaded into ES it appears to work

File: usaspending_api/etl/es_etl_helpers.py [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+        template = '[{time}] [ERROR] {msg}'
+        print_msg = template.format(time=t, msg=msg)
+    else:
+        template = '[{time}] {complex:<20} | {msg}'
+        func = '[' + items.get('f', 'main') + ']'
+        job = items.get('job', None)
+        j = ''
+        if job:
+            j = ' (#{})'.format(job)
+    print_msg = template.format(time=t, complex=func + j, msg=msg)
```
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/etl/es_etl_helpers.py]: good catch

File: usaspending_api/etl/es_etl_helpers.py [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+
+    print(print_msg)
+
+
+def csv_row_count(filename, has_header=True):
```
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/etl/es_etl_helpers.py]: Yeah, I had the same exact reaction. I thought there had to be a better way

File: usaspending_api/etl/es_etl_helpers.py [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+    with open(filename, 'r') as f:
+        row_count = sum(1 for row in csv.reader(f))
+    if has_header:
+        row_count -= 1
+    return row_count
+
+# ==============================================================================
```
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/etl/es_etl_helpers.py]: The reason the "SQL" strings are below is the view SQL was originally included here and is 90 lines. No problem moving to top of file as they are globals

File: usaspending_api/etl/management/commands/es_rapidloader.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        self.config['verbose'] = True if options['verbosity'] > 1 else False
+        self.config['fiscal_years'] = options['fiscal_years']
+        self.config['directory'] = options['dir'] + os.sep
+        self.config['provide_deleted'] = options['deleted']
+        self.config['recreate'] = options['recreate']
+        self.config['stale'] = options['stale']
+        self.config['keep'] = options['keep']
+
+        try:
+            self.config['starting_date'] = date(*[int(x) for x in options['since'].split('-')])
```
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/etl/management/commands/es_rapidloader.py]: If there was I can't think of it now

**Pull request merged**